src/core/coap/coap_server.cpp

```diff
@@ -173,7 +173,7 @@ void Server::ProcessReceivedMessage(Message &aMessage, const Ip6::MessageInfo &a
         break;
     }
 
-    coapOption = header.GetCurrentOption();
+    coapOption = header.GetFirstOption();
```

Comment:
This also fixes a possible bug:
When payload with no option arrives, GetCurrentOption would return non-null value, i.e. pointer to mOption, which, in this case, would be uninitialized and could result in unexpected behavior.
Maybe Init() should be called in FromMessage?